### ISSUE TITLE
Do not imply a factory wide `derived`-annotation to inherited attributes.

### ISSUE THREAD

**Issue by adirla** [ldegen/tbob#13]
current behaviour

    @attr "strasse"
      .type @optional @string
      .fill "In der Pampa"

will fill the attribute when its' input string is empty.

wanted behaviour:
@optional attributes shouldn't be filled. (Unless explicitly stated. Maybe a future feature)

**Comment by ldegen** [comment]
We already have a command line switch for disabling fill strategies, see #8. If it is not working correctly, it would be nice to have a minimal example to reproduce this.

**Comment by ldegen** [comment]
ok, first of all: The behavior of `@optional` is unrelated to whether a fill strategy is applied or not. I do not want to change this. 

Next, and more important: There is a bug, that need to be fixed. Using `-d` (or equivalently `onlyFillDerivedAttributes: true`) should make sure that *all* fill strategies are ignored *unless* the respective attribute is marked as `derived`. Unfortunately, this does not always work as expected:

``` coffee
module.exports = ->
  @factory "Adresse", ->
    @attr "strasse"
      .type @optional @string
      .fill "In der Pampa"

  @factory "AdresseES", ->
    @extend "Adresse"
    @meta derived: true
    @attr "_sort"
      .type @optional @string
      .fill ["strasse"], (s)->s?.toLowerCase()
```

Here, all attributes in factory `AdresseES` are marked as `derived`. But this (currently) includes attributes
inherited from `Adresse`, which is sort-of purpose defeating, and clearly needs to be fixed.

**Comment by ldegen** [comment]
@adirla I took the liberty to update the title to reflect the above statements.

**Comment by ldegen** [comment]
Come to think of it... we are using the  `.fill()` directive for two entirely different purposes. Which is probably a stupid idea to begin with, as @adirla  already mentioned when we last talked about it. 

How about we have to two separate directives instead:

  - `.fill()` will be used to fill in missing attributes with example data for testing purposes -- this is exactly
    as it was originally intended. There would be some switch (like `-d`) to entirely disable *all* fill 
    strategies, without any exceptions.

  - a new directive `.derive()` will be introduced, which works like `.fill()` but will always be used,
    regardless of any switches.

  - ~~Any explicit input for a _derived_ will be rejected and raise an error. If the attribute is derived, it is so *always*. Anything else smells like an error.~~ *UPDATE*: no, wait: With `.fill()`  it is possible to transform an attribute "in place". Does this make sense for `.derive()` as well? 

  - ~~Having an attribute with both `.fill()` and `.derive()` will raise an error. An attribute is either derived or not. Of it is derived, there is no need for a fill strategy.~~ *UPDATE*: same as above: there might be border cases where it is actually desirable to combine both. Need to think about this.

**Comment by ldegen** [comment]
Attributes can now have both a `fill` and a `derive` strategy. Both work very similar but serve different purposes as explained above. There are interesting border cases I already mentioned above: what happens, if a `derive` strategy is given, but the input contains an explicit value for the same attribute?
Per default, we want to reject this input. The attribute is derived, so we do *not* expect explicit values in the input. On the other hand, we follow a convention that we have taken over from Rosie: If an attribute specifies a dependency on itself, its `fill` strategy will get access to the input value specified for that attribute. This is very useful in some situations because it allows to combine / merge derived values with the explicit data received as input. I think it is consistent to support the same convention for `derive` strategies. So the border cases are solved like this:

- An attribute with a `derive` strategy *will* accept explicit input, if and only if the derive strategy specifies 
  the attribute itself as a dependency. 

- An attribute is allowed to have both, a `derive` and a `fill` strategy. In such a case the `fill` strategy is
  applied first and the result is passed on as input to the `derive`-Strategy.
  Since this effectively produces input for the attribute, the first rule also applies: the 
  `derive`-strategy must explicitly specify the attribute itself as a dependency. Otherwise an error
  will be raised.

Most of this is already implemented:

- [x] low-level support for `derive` and `fill` on the `Attribute` level.
- [x] extend the DSL syntax to support `derive` in addition to `fill`.
- [x] deprecate/remove the behaviour introduced in #8. Instead, introduce a switch to disable `fill`-strategies entirely.

**Comment by ldegen** [comment]
For the sake of completeness: The original *technical* problem that caused this issue is that attributes do not track where they are originally defined. So in our example the attribute `strasse` inherits the annotation `derived: true` from the factory `AdresseES`, although it was originally defined in `Adresse`. 

This "inheriting"-behavior seemed fine when I implemented it, and I still like it, because it is dead simple.
For most annotations this behavior is sufficient or even desirable,  but for something like `derived` it makes no sense, as we have seen. 

Since we use separate strategies for `fill` and `derive` in the future, we will have no further need for the `derived` annotation, so that should be a non-issue, and it will not be addressed for now.

Still, if you think we need a more „subtle“ treatment meta-data and inheritance, please feel free to open another issue.

**Comment by ldegen** [comment]
The old `-d` (which is equivalently to passing `onlyFillDerivedAttributes: true` through the API) will now produce a deprecation warning.

I introduced a new switch `-n` or `--no-fill` which is equivalent to passing `disableFillStrategies: true` through the API. This completely disables *all* `fill` strategies, but still uses the `derive` strategies.

**Comment by ldegen** [comment]
landed in tbob@0.11.0
